blockchain_storage/src/mock.rs

```diff
@@ -53,9 +50,11 @@ mockall::mock! {
         fn del_block_id_at_height(&mut self, height: &BlockHeight) -> crate::Result<()>;
     }
 
-    impl storage::Transactional<'_> for Store {
-        type Transaction = MockStoreTx;
-        fn start_transaction(&mut self) -> MockStoreTx;
+    impl storage::traits::Transactional<'_> for Store {
+        type TransactionRo = MockStoreTx;
+        type TransactionRw = MockStoreTx;
+        fn start_transaction_ro(&self) -> MockStoreTx;
+        fn start_transaction_rw(&self) -> MockStoreTx;
```

Comment:
Are you sure this is correct, taking a `&self` even though it's r/w?

Reply:
This follows the [interior mutablility pattern](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html), same as `Mutex`or `LockRw`, or `RefCell`. The basic storage implementation uses a `RwLock` under the hood in this iteration. If you look for example at [RefCell doc](https://doc.rust-lang.org/std/cell/struct.RefCell.html), notice that both `borrow` and `borrow_mut` take `&self` wihtout mut.

There are basically three similar methods, taking `RefCell` as an example again:

1. `borrow(&self) -> Ref<'_, T>`
2. `borrow_mut(&self) -> RefMut<'_, T>`
3. `get_mut(&mut self) -> &mut T`

The `get_mut` method guarantees (due its type signature) that the RefCell is the only reference to the data. It is only usable upon initialization, once you start sharing the cell, you have to use `borrow` and `borrow_mut` which provide a runtime mechanism to enforce the aliasing rule. With storage, it is similar situation. The aliasing rules are enforced by a mutex internally so the storage can be shared by multiple subsystems.

In the real implementation (non-mock) the mutable and immutable case are differentiated by the type of transaction that is returned, i.e. `start_transaction_ro` returns a `Transactional::TransactionRo`, whereas `start_transaction_rw` returns a `Transactional::TransactionRw`, These two are analogous to (1) and (2). With mock I took the liberty to take a shortcut and simplify things a bit by having just one transaction type representing both mutable and immutable transactions.

Reply:
Well, nevertheless, I'm wondering whether it nominally makes more sense to make it `&mut self` just for the meaningfulness of it. I mean, we're using the rust mutability properties to map them to database operation mutability properties, and we don't need to do that anyway, but it's for our own pleasure and safety in our code. So, it's going through my head whether we should just make it `&mut self` because we're writing, to distinguish from reading. Just like we make `get` a `&self` while a `set` is `&mut self`.

Reply:
Turns out having `transaction_rw` take `&mut self` is not nicely doable until blockchain storage also uses two separate types for read-only vs. read-write transactions. I.e. it makes it difficult to have one transaction type that implements both `TransactionRw` functionalisy and `TransactionRo` functionality.